### PR TITLE
resilience to more types of failure

### PR DESCRIPTION
**Pull request by lwwmanning**
## Before this PR

Potential filesystem conflicts necessitated bumping the experimentName before every new run, and runs would fail due to a single transient failure (e.g., during validation).

## After this PR
==COMMIT_MSG==
Attempt to record as much useful information as possible from runs, and reduce transient errors. We accomplish this by: using a unique filesystem paths for all outputs, recording verification as a column rather than throwing on failure, and retrying failed queries.
==COMMIT_MSG==

## Possible downsides?

If failures are expensive, the retries will make the number of failures (and thus time to total failure, data written to storage, etc.) greater.

### CODE REVIEW

File: spark-tpcds-benchmark-runner/build.gradle [hunk 1]
```diff
@@ -24,12 +24,16 @@ dependencies {
...
     implementation 'com.fasterxml.jackson.datatype:jackson-datatype-guava'
+    implementation 'com.github.rholder:guava-retrying'
     implementation 'com.google.guava:guava'
     implementation 'com.palantir.common:streams'
     implementation 'com.palantir.safe-logging:preconditions'
     implementation 'com.palantir.conjure.java.runtime:client-config'
     implementation 'org.apache.spark:spark-sql_2.12'
     implementation 'org.apache.spark:spark-hadoop-cloud_2.12'
+    implementation('org.mapdb:mapdb') {
+        exclude group: 'net.jpountz.lz4', module: 'lz4'
```
Comment (rahij) [inline spark-tpcds-benchmark-runner/build.gradle:35]: curious, why is this necessary?

File: spark-tpcds-benchmark-runner/src/main/java/com/palantir/spark/benchmark/paths/BenchmarkPaths.java [hunk 2]
```diff
@@ -59,13 +60,15 @@ public String parquetDir(int scale) {
...
+    public String experimentResultLocation(QuerySessionIdentifier identifier, int attempt) {
         return String.join(
                 File.separator,
                 "experiments_query_results",
                 experimentName,
-                String.format("scale=%d", scale),
-                StringUtils.removeEnd(queryName, ".sql"));
+                String.format("scale=%d", identifier.scale()),
+                StringUtils.removeEnd(identifier.queryName(), ".sql"),
+                String.format("session=%s", identifier.session()),
```
Comment (rahij) [inline spark-tpcds-benchmark-runner/src/main/java/com/palantir/spark/benchmark/paths/BenchmarkPaths.java:70]: should we do the same for experimentCorrectnessHashesRoot? (I'm still not sure where the correctness hashes are actually from :D)

File: spark-tpcds-benchmark-runner/build.gradle (hunk 1 above)
Comment (lwwmanning) [reply spark-tpcds-benchmark-runner/build.gradle:35]: this library changed its maven coordinate to `org.lz4:lz4-java` in a later release, which spark depends on; having both on the classpath caused problems. See https://github.com/palantir/spark-tpcds-benchmark/blob/e49806fd0ee5cf0e93b15d64d386d5d3c5b8f470/versions.lock#L213

File: spark-tpcds-benchmark-runner/src/main/java/com/palantir/spark/benchmark/paths/BenchmarkPaths.java (hunk 2 above)
Comment (lwwmanning) [reply spark-tpcds-benchmark-runner/src/main/java/com/palantir/spark/benchmark/paths/BenchmarkPaths.java:70]: we could add `session`... from my understanding of the code, we just write the table hash on the first iteration, so on subsequent iterations we compare for equality (or like, if we pick our paths so that they're the same across runs, then it gets written on first iteration of first run and read from there forever after)

**Comment by rahij** [comment]
👍
